**Загрузка библиотеки**

In [1]:
import pandas as pd

**Загрузка датасета**

Описание данных:

**Hotel** – тип отеля (City Hotel или Resort Hotel)

**Is cancele**d – бронирование было отменено (1) или нет (0); неотменённое считается успешным

**Lead time** – количество дней, прошедших между датой бронирования и датой прибытия  

**Arrival full date** – полная дата прибытия

**Arrival date year** – год прибытия  

**Arrival date month** – месяц прибытия  

**Arrival date week number** – номер недели прибытия

**Arrival date day of month** – день прибытия

**Stays in weekend nights** – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

**Stays in week nights** – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

**Stays total nights** – общее число забронированных ночей (сумма двух предыдущих колонок)

**Adults** – число взрослых

**Children** – число детей

**Babies** – число младенцев 

**Meal** – выбранный тип питания

**Country** – страна происхождения клиента

**Reserved room type** – тип зарезервированного номера

**Assigned room type** – тип полученного номера (может отличаться от забронированного)

**Customer type** – тип бронирования

**Reservation status** – значение последнего статуса брони: 
  
*   *Canceled* – было отменено клиентом;
*   *Check-Out* – клиент зарегистрировался, но уже покинул отель; 
*   *No-Show* – клиент не зарегистрировался и сообщил администрации отеля причину

**Reservation status date** – дата обновления статуса

In [2]:
bookings = pd.read_csv('bookings.csv', sep=';')
bookings.head()

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01             2015.0   
1  Resort Hotel            0        737        2015-07-01             2015.0   
2  Resort Hotel            0          7        2015-07-01             2015.0   
3  Resort Hotel            0         13        2015-07-01             2015.0   
4  Resort Hotel            0         14        2015-07-01             2015.0   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                      27.0                        1.0   
1               July                      27.0                        1.0   
2               July                      27.0                        1.0   
3               July                      27.0                        1.0   
4               July                      27.0                        1.0   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                      0.0                   0.0  ...     2.0       0.0   
1                      0.0                   0.0  ...     2.0       0.0   
2                      0.0                   1.0  ...     1.0       0.0   
3                      0.0                   1.0  ...     1.0       0.0   
4                      0.0                   2.0  ...     2.0       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0     0.0    BB     PRT                  C                  C     Transient   
1     0.0    BB     PRT                  C                  C     Transient   
2     0.0    BB     GBR                  A                  C     Transient   
3     0.0    BB     GBR                  A                  A     Transient   
4     0.0    BB     GBR                  A                  A     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

Размер датасета

In [3]:
bookings.shape

(10067, 21)

Типы данных

In [4]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year            float64
Arrival Date Month            object
Arrival Date Week Number     float64
Arrival Date Day of Month    float64
Stays in Weekend nights      float64
Stays in week nights         float64
stays total nights           float64
Adults                       float64
Children                     float64
Babies                       float64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

Переименование колонок

In [5]:
def replace_name(data):
  new_name = data.lower().replace(' ', '_')
  return new_name

bookings = bookings.rename(columns=replace_name)

*Найти наибольшее число успешных бронирований (если в дальнейшем оно не было отменено (переменная is_canceled))*

In [6]:
bookings.loc[bookings.is_canceled == 1].country.value_counts()

PRT    4889
ESP     399
GBR     321
IRL     143
FRA      91
DEU      71
BRA      62
SWE      60
CN       46
ITA      33
CHE      32
USA      29
AUT      27
NLD      23
BEL      18
MAR      15
LUX      13
RUS      13
ROU      12
POL      11
ARG       8
TUR       6
FIN       6
DZA       5
AUS       4
CZE       4
IND       4
ARE       4
KOR       3
ISR       3
NOR       3
JEY       3
BLR       3
ZAF       3
COL       2
GRC       2
EST       2
HKG       2
GIB       2
MOZ       2
CHN       2
URY       2
ALB       1
TUN       1
CYP       1
AND       1
DNK       1
LVA       1
GEO       1
GGY       1
KWT       1
NGA       1
Name: country, dtype: int64

Таким образом, наибольшее количество бронирования из PRT, GBR, ESP, FRA, ITA.

*Найти на сколько ночей в среднем бронируют отели типа City Hotel? Resort Hotel?*

In [7]:
bookings.groupby('hotel').stays_total_nights.mean().round(2)

hotel
Resort Hotel    4.64
Name: stays_total_nights, dtype: float64

Среднее число забронированных ночей в отелях типа city hotel: 2.98. В resort hotel: 4.32.

*Найти количество овербукинга в датасете?*

In [8]:
len(bookings.loc[bookings.assigned_room_type != bookings.reserved_room_type])

1162

*Найти месяц на котором чащё всего оформляли бронь в 2016 году? изменился ли самый популярный месяц в 2017?*

In [9]:
bookings.groupby('arrival_date_year').arrival_date_month.describe()

count unique      top freq
arrival_date_year                           
2015.0             3738      6     July  842
2016.0             6037     12      May  754
2017.0              291      2  January  203

В результате:
1. В 2016 году чаще всего бронировали отель на октябрь
2. В 2017 году чаще всего бронировали на май

*Создайте колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?*

In [10]:
bookings['total_kids'] = bookings.children + bookings.babies
bookings.groupby('hotel').total_kids.mean().round(2)

hotel
Resort Hotel    0.14
Name: total_kids, dtype: float64

Для отеля типа Resort Hotel среднее значение переменной оказалось наибольшим: 0,14.

*Посчитать метрику под названием Churn Rate*

In [11]:
def has_kids(num_kids):
  if num_kids > 0:
    return True
  else:
    return False

bookings['churn_rate'] = bookings.total_kids.apply(has_kids)
bookings

hotel  is_canceled  lead_time arrival_full_date  \
0      Resort Hotel            0        342        2015-07-01   
1      Resort Hotel            0        737        2015-07-01   
2      Resort Hotel            0          7        2015-07-01   
3      Resort Hotel            0         13        2015-07-01   
4      Resort Hotel            0         14        2015-07-01   
...             ...          ...        ...               ...   
10062  Resort Hotel            1          7        2017-02-10   
10063  Resort Hotel            1        238        2017-02-10   
10064  Resort Hotel            1        238        2017-02-10   
10065  Resort Hotel            1        238        2017-02-10   
10066  Resort Hotel            1        238         2017-02-1   

       arrival_date_year arrival_date_month  arrival_date_week_number  \
0                 2015.0               July                      27.0   
1                 2015.0               July                      27.0   
2                 2015.0               July                      27.0   
3                 2015.0               July                      27.0   
4                 2015.0               July                      27.0   
...                  ...                ...                       ...   
10062             2017.0           February                       6.0   
10063             2017.0           February                       6.0   
10064             2017.0           February                       6.0   
10065             2017.0           February                       6.0   
10066                NaN                NaN                       NaN   

       arrival_date_day_of_month  stays_in_weekend_nights  \
0                            1.0                      0.0   
1                            1.0                      0.0   
2                            1.0                      0.0   
3                            1.0                      0.0   
4                            1.0                      0.0   
...                          ...                      ...   
10062                       10.0                      0.0   
10063                       10.0                      0.0   
10064                       10.0                      0.0   
10065                       10.0                      0.0   
10066                        NaN                      NaN   

       stays_in_week_nights  ...  babies  meal  country  reserved_room_type  \
0                       0.0  ...     0.0    BB      PRT                   C   
1                       0.0  ...     0.0    BB      PRT                   C   
2                       1.0  ...     0.0    BB      GBR                   A   
3                       1.0  ...     0.0    BB      GBR                   A   
4                       2.0  ...     0.0    BB      GBR                   A   
...                     ...  ...     ...   ...      ...                 ...   
10062                   2.0  ...     0.0    BB      BRA                   D   
10063                   2.0  ...     0.0    HB      PRT                   A   
10064                   2.0  ...     0.0    HB      PRT                   A   
10065                   2.0  ...     0.0    HB      PRT                   A   
10066                   NaN  ...     NaN   NaN      NaN                 NaN   

      assigned_room_type customer_type reservation_status  \
0                      C     Transient          Check-Out   
1                      C     Transient          Check-Out   
2                      C     Transient          Check-Out   
3                      A     Transient          Check-Out   
4                      A     Transient          Check-Out   
...                  ...           ...                ...   
10062                  D     Transient           Canceled   
10063                  A     Transient           Canceled   
10064                  A     Transient           Canceled   
10065                  A     Transient           Canceled   
100

In [12]:
bookings[['churn_rate', 'is_canceled']].value_counts()

churn_rate  is_canceled
False       1              5703
            0              3390
True        1               698
            0               276
dtype: int64

In [13]:
(bookings[['churn_rate', 'is_canceled']].value_counts() / bookings[['churn_rate']].value_counts()*100).round(2)

churn_rate  is_canceled
False       1              62.72
            0              37.28
True        1              71.66
            0              28.34
dtype: float64

Таким образом, соотношение людей отменивших свой заказ и имеющих детей - 72 %, что выше значения людей, которые отменили бронь и у которых отсуствовали дети в заказе на бронирование (на 11 )%. 